File: engine/typescript/src/world/world.ts
// world/world.ts
//
// The world model. Owns all rooms and objects, answers scope queries,
// and handles player movement.
//
// Populated at startup by loadWorld() / World.load(). Never hardcodes game
// content — all data comes from JSON via the loader.

import type { Connector, GameObject, Room, WorldContext } from '../types.ts'

// ---------------------------------------------------------------------------
// Module-level state — populated by World.load()
// ---------------------------------------------------------------------------
let rooms:          Record<string, Room>       = {}
let objects:        Record<string, GameObject> = {}
let currentRoomKey  = ''
let startRoomKey    = ''

interface ObjectSnap { location: string | null; locked?: boolean; isOpen?: boolean; moved?: boolean }
interface RoomSnap   { visited: boolean }
const initialState: Record<string, ObjectSnap | RoomSnap> = {}

// ---------------------------------------------------------------------------
// Internal compositor helpers
// ---------------------------------------------------------------------------

function isIlluminated(room: Room): boolean {
    return room.isLit !== false
}

function unmentionAll(): void {
    for (const obj of Object.values(objects)) {
        obj.mentioned = false
    }
}

function hasActiveSpecialDesc(obj: GameObject): boolean {
    if (obj.initSpecialDesc && !obj.moved) return true
    if (obj.specialDesc) return true
    return false
}

function showSpecialDesc(obj: GameObject): string | null {
    let text: string | ((self: GameObject) => string) | undefined
    if (obj.initSpecialDesc && !obj.moved) {
        text = obj.initSpecialDesc
    } else {
        text = obj.specialDesc
    }
    if (!text) return null
    const result = typeof text === 'function' ? text(obj) : text
    obj.mentioned = true
    return result
}

function article(name: string): string {
    return /^[aeiou]/i.test(name) ? 'an' : 'a'
}

function buildMiscSentence(items: GameObject[]): string {
    const names = items.map(obj => {
        obj.mentioned = true
        return `${article(obj.name)} ${obj.name}`
    })
    if (names.length === 1) return `You can also see: ${names[0]}.`
    if (names.length === 2) return `You can also see: ${names[0]} and ${names[1]}.`
    const last = names.pop()!
    return `You can also see: ${names.join(', ')}, and ${last}.`
}

function listContents(room: Room, ctx: Required<Pick<WorldContext, 'excluded'>>): string | null {
    const firstSpecial:  GameObject[] = []
    const miscItems:     GameObject[] = []
    const secondSpecial: GameObject[] = []

    for (const key of room.objects) {
        const obj = objects[key]
        if (!obj || ctx.excluded[key] || obj.mentioned) continue
        if (obj.location !== currentRoomKey) continue

        if (hasActiveSpecialDesc(obj)) {
            if (obj.specialDescBeforeContents !== false) {
                firstSpecial.push(obj)
            } else {
                secondSpecial.push(obj)
            }
        } else if (obj.listed !== false && !obj.scenery
                   && !obj.remapIn && !obj.remapOn
                   && !(obj.contType === 'in' && obj.isOpen)) {
            miscItems.push(obj)
        }
    }

    miscItems.sort((a, b) => a.name.localeCompare(b.name))

    const byOrder = (a: GameObject, b: GameObject) =>
        (a.specialDescOrder ?? 100) - (b.specialDescOrder ?? 100)
    firstSpecial.sort(byOrder)
    secondSpecial.sort(byOrder)

    const result: string[] = []

    for (const obj of firstSpecial) {
        const text = showSpecialDesc(obj)
        if (text) result.push(text)
    }

    if (miscItems.length > 0) {
        result.push(buildMiscSentence(miscItems))
    }

    for (const obj of secondSpecial) {
        const text = showSpecialDesc(obj)
        if (text) result.push(text)
    }

    // Stage 4A: Composite objects (remapIn / remapOn) get a dedicated paragraph.
    // They are excluded from the misc sentence and described here instead.
    for (const key of room.objects) {
        const obj = objects[key]
        if (!obj || ctx.excluded[key] || obj.mentioned) continue
        if (obj.location !== currentRoomKey) continue
        if (!obj.remapIn && !obj.remapOn) continue

        const lines: string[] = [`There is a ${obj.name} here.`]

        if (obj.remapOn) {
            const sub = World.getObject(obj.remapOn)
            if (sub) {
                const names = World.contentsOf(sub._key!)
                    .filter(item => !item.mentioned)
                    .map(item => { item.mentioned = true; return item.name })
                if (names.length > 0) {
                    lines.push(`On the ${sub.name}: ${names.join(', ')}.`)
                }
                sub.mentioned = true
            }
        }

        if (obj.remapIn) {
            const sub = World.getObject(obj.remapIn)
            if (sub) {
                if (sub.isOpen) {
                    const names = World.contentsOf(sub._key!)
                        .filter(item => !item.mentioned)
                        .map(item => { item.mentioned = true; return item.name })
                    if (names.length > 0) {
                        lines.push(`The ${sub.name} is open. It contains: ${names.join(', ')}.`)
                    } else {
                        lines.push(`The ${sub.name} is open and empty.`)
                    }
                }
                sub.mentioned = true
            }
        }

        obj.mentioned = true
        result.push(lines.join(' '))
    }

    // Stage 4B: Direct containers in the room (on-surfaces and open in-containers).
    for (const key of room.objects) {
        const cont = objects[key]
        if (!cont || !cont.contType || ctx.excluded[key] || cont.mentioned) continue
        if (cont.location !== currentRoomKey) continue

        if (cont.contType === 'on') {
            const names = World.contentsOf(key)
                .filter(item => !item.mentioned)
                .map(item => { item.mentioned = true; return item.name })
            if (names.length > 0) {
                result.push(`On the ${cont.name}: ${names.join(', ')}.`)
            }
            cont.mentioned = true
        } else if (cont.contType === 'in' && cont.isOpen) {
            const names = World.contentsOf(key)
                .filter(item => !item.mentioned)
                .map(item => { item.mentioned = true; return item.name })
            if (names.length > 0) {
                result.push(`The ${cont.name} is open. It contains: ${names.join(', ')}.`)
            } else {
                result.push(`The ${cont.name} is open and empty.`)
            }
            cont.mentioned = true
        }
        // Closed in-containers: skip (in misc list or invisible if listed:false).
    }

    return result.length > 0 ? result.join('\n\n') : null
}

function listExits(room: Room): string | null {
    const dirs = Object.keys(room.exits)
    if (dirs.length === 0) return null
    dirs.sort()
    return 'Exits: ' + dirs.join(', ') + '.'
}

// ---------------------------------------------------------------------------
// World API
// ---------------------------------------------------------------------------
export const World = {

    load(
        roomsTable:   Record<string, Room>,
        objectsTable: Record<string, GameObject>,
        startRoom:    string
    ): void {
        rooms          = roomsTable
        objects        = objectsTable
        startRoomKey   = startRoom
        currentRoomKey = startRoom

        for (const [key, obj] of Object.entries(objects)) {
            obj._key = key
            const snap: ObjectSnap = { location: obj.location }
            if (obj.locked  !== undefined) snap.locked  = obj.locked
            if (obj.isOpen  !== undefined) snap.isOpen  = obj.isOpen
            if (obj.moved   !== undefined) snap.moved   = obj.moved
            initialState[key] = snap
        }
        for (const key of Object.keys(rooms)) {
            initialState[key] = { visited: false }
        }

        // Build room.objects (direct children only) from object location properties.
        for (const room of Object.values(rooms)) room.objects = []
        for (const [objKey, obj] of Object.entries(objects)) {
            const loc = obj.location
            if (loc && rooms[loc]) rooms[loc].objects.push(objKey)
        }
        for (const room of Object.values(rooms)) room.objects.sort()
    },

    currentRoom(): Room {
        return rooms[currentRoomKey]
    },

    currentRoomKey(): string {
        return currentRoomKey
    },

    currentContext(): WorldContext {
        return {}
    },

    inScope(): GameObject[] {
        const scope: GameObject[] = []
        const room = rooms[currentRoomKey]

        // Recursively add contents of an object container, sorted by key.
        function addContainerContents(contKey: string): void {
            const children = Object.values(objects)
                .filter(o => o.location === contKey)
                .sort((a, b) => (a._key ?? '').localeCompare(b._key ?? ''))
            for (const obj of children) {
                scope.push(obj)
                if (obj.contType === 'on'
                        || (obj.contType === 'in' && obj.isOpen)) {
                    addContainerContents(obj._key!)
                }
            }
        }

        // Top-level room objects (room.objects is kept sorted by key).
        // In an unlit room, only objects with visibleInDark = true are in scope.
        const dark = !isIlluminated(room)
        for (const key of room.objects) {
            const obj = objects[key]
            if (obj && (!dark || obj.visibleInDark)) {
                scope.push(obj)
                if (obj.contType === 'on'
                        || (obj.contType === 'in' && obj.isOpen)) {
                    addContainerContents(obj._key!)
                }
            }
        }

        // Inventory (sorted by key for determinism).
        const inv = Object.values(objects)
            .filter(o => o.location === 'inventory')
            .sort((a, b) => (a._key ?? '').localeCompare(b._key ?? ''))
        for (const obj of inv) scope.push(obj)

        return scope
    },

    contentsOf(objKey: string): GameObject[] {
        const result = Object.values(objects).filter(o => o.location === objKey)
        result.sort((a, b) => a.name.localeCompare(b.name))
        return result
    },

    resolveContainer(obj: GameObject, prep: string): GameObject {
        if (prep === 'in' && obj.remapIn) return objects[obj.remapIn]!
        if (prep === 'on' && obj.remapOn) return objects[obj.remapOn]!
        return obj
    },

    describeCurrentRoom(): string {
        const room = rooms[currentRoomKey]

        // Step 1: Reset mentioned flags.
        unmentionAll()

        const parts: string[] = []

        // Step 2: Room title.
        parts.push(isIlluminated(room) ? room.name : (room.darkName ?? 'In the dark'))

        // Step 3: Dark branch.
        if (!isIlluminated(room)) {
            const raw = room.darkDesc ?? "It is pitch black; you can't see a thing."
            parts.push(typeof raw === 'function' ? raw(room) : raw)
            room.visited = true
            return parts.join('\n')
        }

        // Step 4: Build room context.
        const excluded: Record<string, boolean> = {}
        const ctx: WorldContext = {
            firstVisit: !room.visited,
            excluded,
            exclude: (key: string) => { excluded[key] = true },
        }

        // Step 5: Room description body.
        // Title and body join with a single newline; subsequent blocks use double.
        let out = parts[0] + '\n' + room.description(room, ctx)

        // Steps 6–7: Object listing and exit listing.
        if (!room.suppressListing) {
            const listing = listContents(room, { excluded })
            if (listing) out += '\n\n' + listing

            const exits = listExits(room)
            if (exits) out += '\n\n' + exits
        }

        // Step 8: Mark as visited.
        room.visited = true

        return out
    },

    describeInventory(): string {
        const carried = Object.values(objects)
            .filter(o => o.location === 'inventory')
            .map(o => o.name)
            .sort()
        if (carried.length === 0) return 'You are carrying nothing.'
        return 'You are carrying: ' + carried.join(', ') + '.'
    },

    reset(): void {
        for (const [key, snap] of Object.entries(initialState)) {
            if (key in rooms) {
                rooms[key].visited = false
            } else if (key in objects) {
                const s = snap as ObjectSnap
                objects[key].location = s.location
                if (s.locked !== undefined) objects[key].locked = s.locked
                if (s.isOpen !== undefined) objects[key].isOpen = s.isOpen
                objects[key].moved = s.moved
            }
        }
        currentRoomKey = startRoomKey

        // Rebuild room.objects from restored object locations.
        for (const room of Object.values(rooms)) room.objects = []
        for (const [objKey, obj] of Object.entries(objects)) {
            const loc = obj.location
            if (loc && rooms[loc]) rooms[loc].objects.push(objKey)
        }
        for (const room of Object.values(rooms)) room.objects.sort()
    },

    moveTo(roomKey: string): void {
        currentRoomKey = roomKey
    },

    moveObject(obj: GameObject, location: string | null): void {
        if (location === 'inventory' && obj.location !== 'inventory') {
            obj.moved = true
        }

        // Remove from old room's direct-child list if it was directly in a room.
        const oldLoc = obj.location
        if (oldLoc && rooms[oldLoc]) {
            const oldObjs = rooms[oldLoc].objects
            const idx = oldObjs.indexOf(obj._key!)
            if (idx !== -1) oldObjs.splice(idx, 1)
        }

        obj.location = location

        // Add to new room's direct-child list if moving directly into a room.
        if (location && rooms[location]) {
            rooms[location].objects.push(obj._key!)
        }
    },

    getObject(key: string): GameObject | undefined {
        return objects[key]
    },

    getConnector(room: Room, dir: string): Connector | undefined {
        return room.exits[dir]
    },
}

File: engine/lua/world/world.lua
-- world/world.lua
--
-- The world model. Owns all rooms and objects, answers scope queries,
-- and handles player movement.
--
-- Populated at startup by Loader.load() / World.load(). Never hardcodes game
-- content — all data comes from JSON via the loader.
--
-- Other modules never access rooms or objects directly. They always go
-- through World's functions. This keeps the world internals replaceable.

local World = {}

-- ---------------------------------------------------------------------------
-- Module-level state — populated by World.load()
-- ---------------------------------------------------------------------------
local rooms          = {}
local objects        = {}
local currentRoomKey = ""
local startRoomKey   = ""
local helpData       = { default = nil, topics = {} }

-- Snapshot of mutable object/room state taken at load time, used by reset().
local initialState = {}

-- ---------------------------------------------------------------------------
-- World.load — called once by the loader after parsing JSON.
-- Stores the tables and computes the reset snapshot.
-- ---------------------------------------------------------------------------
function World.load(roomsTable, objectsTable, startRoom)
    rooms          = roomsTable
    objects        = objectsTable
    startRoomKey   = startRoom
    currentRoomKey = startRoom

    -- Set _key on every object and take the mutable-state snapshot.
    initialState = {}
    for key, obj in pairs(objects) do
        obj._key = key
        initialState[key] = {
            location = obj.location,
            locked   = obj.locked,
            moved    = obj.moved,
            isOpen   = obj.isOpen,
        }
    end
    for key in pairs(rooms) do
        initialState[key] = { visited = false }
    end

    -- Build room.objects (direct children only) from object location properties.
    for _, room in pairs(rooms) do
        room.objects = {}
    end
    for objKey, obj in pairs(objects) do
        local loc = obj.location
        if loc and rooms[loc] then
            rooms[loc].objects[#rooms[loc].objects + 1] = objKey
        end
    end
    for _, room in pairs(rooms) do
        table.sort(room.objects)
    end

end

-- ---------------------------------------------------------------------------
-- World API — basic queries
-- ---------------------------------------------------------------------------

function World.currentRoom()
    return rooms[currentRoomKey]
end

function World.currentRoomKey()
    return currentRoomKey
end

-- Returns a context table for object description functions.
-- Room descriptions get a richer context built inside describeCurrentRoom().
function World.currentContext()
    return {}
end

-- Returns true if the room has ambient light.
-- Default: lit (isLit == nil or isLit == true).
-- isLit may be a boolean or a compiled condition closure (from a flagCheck).
local function isIlluminated(room)
    if type(room.isLit) == "function" then return room.isLit() end
    return room.isLit ~= false
end

-- Returns all objects currently in scope.
-- Uses sorted room.objects for the top level; recurses into open containers
-- (sorting their children by key for determinism); appends inventory last.
function World.inScope()
    local scope = {}
    local room  = rooms[currentRoomKey]

    -- Recursively add contents of an object container (not used for the room
    -- itself, since room.objects already provides the sorted top-level list).
    local function addContainerContents(contKey)
        local children = {}
        for _, obj in pairs(objects) do
            if obj.location == contKey then
                children[#children + 1] = obj
            end
        end
        table.sort(children, function(a, b) return (a._key or "") < (b._key or "") end)
        for _, obj in ipairs(children) do
            scope[#scope + 1] = obj
            if obj.contType == "on"
                    or (obj.contType == "in" and obj.isOpen) then
                addContainerContents(obj._key)
            end
        end
    end

    -- Top-level room objects (room.objects is kept sorted by key).
    -- In an unlit room, only objects with visibleInDark = true are in scope.
    local dark = not isIlluminated(room)
    for _, key in ipairs(room.objects) do
        local obj = objects[key]
        if obj and (not dark or obj.visibleInDark) then
            scope[#scope + 1] = obj
            if obj.contType == "on"
                    or (obj.contType == "in" and obj.isOpen) then
                addContainerContents(obj._key)
            end
        end
    end

    -- Inventory (sorted by key for determinism).
    local inv = {}
    for _, obj in pairs(objects) do
        if obj.location == "inventory" then
            inv[#inv + 1] = obj
        end
    end
    table.sort(inv, function(a, b) return (a._key or "") < (b._key or "") end)
    for _, obj in ipairs(inv) do
        scope[#scope + 1] = obj
    end

    return scope
end

-- Returns objects directly inside a container, sorted by name.
function World.contentsOf(objKey)
    local result = {}
    for _, obj in pairs(objects) do
        if obj.location == objKey then
            result[#result + 1] = obj
        end
    end
    table.sort(result, function(a, b) return a.name < b.name end)
    return result
end

-- Follows remapIn/remapOn on obj to find the actual container for a PUT operation.
-- If obj has no remap for the given prep, returns obj itself.
function World.resolveContainer(obj, prep)
    if prep == "in" and obj.remapIn then
        return objects[obj.remapIn]
    end
    if prep == "on" and obj.remapOn then
        return objects[obj.remapOn]
    end
    return obj
end

function World.describeInventory()
    local carried = {}
    for _, obj in pairs(objects) do
        if obj.location == "inventory" then
            carried[#carried + 1] = obj.name
        end
    end
    if #carried == 0 then return "You are carrying nothing." end
    table.sort(carried)
    return "You are carrying: " .. table.concat(carried, ", ") .. "."
end

-- ---------------------------------------------------------------------------
-- World.loadHelp — stores help content loaded from events.json.
-- data: { default = "...", topics = { examine = "...", ... } }
-- ---------------------------------------------------------------------------
function World.loadHelp(data)
    helpData.default = data.default
    helpData.topics  = data.topics or {}
end

-- ---------------------------------------------------------------------------
-- World.getHelp — returns help text for a topic string, or the default text.
-- topic: a lowercased string from intent.dobjWords, e.g. "examine" or "".
-- Falls back to helpData.default, then a bare engine fallback.
-- ---------------------------------------------------------------------------
function World.getHelp(topic)
    if topic and topic ~= "" then
        local text = helpData.topics[topic]
        if text then return text end
        return "No help available for '" .. topic .. "'."
    end
    return helpData.default or "Type commands to interact with the world. Try HELP [topic] for specifics."
end

function World.reset()
    for key, snap in pairs(initialState) do
        if rooms[key] then
            rooms[key].visited = false
        elseif objects[key] then
            objects[key].location = snap.location
            if snap.locked ~= nil then
                objects[key].locked = snap.locked
            end
            if snap.isOpen ~= nil then
                objects[key].isOpen = snap.isOpen
            end
            objects[key].moved = snap.moved
        end
    end
    currentRoomKey = startRoomKey

    -- Rebuild room.objects from restored object locations.
    for _, room in pairs(rooms) do
        room.objects = {}
    end
    for objKey, obj in pairs(objects) do
        local loc = obj.location
        if loc and rooms[loc] then
            rooms[loc].objects[#rooms[loc].objects + 1] = objKey
        end
    end
    for _, room in pairs(rooms) do
        table.sort(room.objects)
    end
end

function World.moveTo(roomKey)
    currentRoomKey = roomKey
end

function World.moveObject(obj, location)
    -- Track that this object has been moved by the player (affects initSpecialDesc).
    if location == "inventory" and obj.location ~= "inventory" then
        obj.moved = true
    end

    -- Remove from old room's direct-child list if it was directly in a room.
    local oldLoc = obj.location
    if oldLoc and rooms[oldLoc] then
        local oldObjs = rooms[oldLoc].objects
        for i, key in ipairs(oldObjs) do
            if key == obj._key then
                table.remove(oldObjs, i)
                break
            end
        end
    end

    obj.location = location

    -- Add to new room's direct-child list if moving directly into a room.
    if location and rooms[location] then
        local newObjs = rooms[location].objects
        newObjs[#newObjs + 1] = obj._key
    end
end

function World.getObject(key)
    return objects[key]
end

function World.getConnector(room, dir)
    return room.exits[dir]
end

-- ---------------------------------------------------------------------------
-- Room description compositor
-- ---------------------------------------------------------------------------

-- Clears the mentioned flag on every object before each LOOK.
local function unmentionAll()
    for _, obj in pairs(objects) do
        obj.mentioned = false
    end
end

-- Returns true if obj has an active specialDesc or initSpecialDesc paragraph.
local function hasActiveSpecialDesc(obj)
    if obj.initSpecialDesc and not obj.moved then return true end
    if obj.specialDesc then
        -- If specialDesc is a function (e.g. conditional array form), evaluate it
        -- to check if it currently produces any text.
        if type(obj.specialDesc) == "function" then
            return obj.specialDesc(obj) ~= ""
        end
        return true
    end
    return false
end

-- Renders an object's specialDesc paragraph and marks it as mentioned.
local function showSpecialDesc(obj)
    local text
    if obj.initSpecialDesc and not obj.moved then
        text = obj.initSpecialDesc
    else
        text = obj.specialDesc
    end
    if not text then return nil end
    if type(text) == "function" then text = text(obj) end
    if not text or text == "" then return nil end
    obj.mentioned = true
    return text
end

-- Returns "an" before vowel-initial words, "a" otherwise.
local function article(name)
    return name:match("^[aeiouAEIOU]") and "an" or "a"
end

-- Builds the "You can also see: X, Y, and Z." sentence for unlisted misc items.
local function buildMiscSentence(items)
    local names = {}
    for _, obj in ipairs(items) do
        names[#names + 1] = article(obj.name) .. " " .. obj.name
        obj.mentioned = true
    end
    if #names == 1 then
        return "You can also see: " .. names[1] .. "."
    elseif #names == 2 then
        return "You can also see: " .. names[1] .. " and " .. names[2] .. "."
    end
    local last = table.remove(names)
    return "You can also see: " .. table.concat(names, ", ") .. ", and " .. last .. "."
end

-- Runs the four-stage listing and returns the assembled string, or nil.
local function listContents(room, ctx)
    local firstSpecial  = {}
    local miscItems     = {}
    local secondSpecial = {}

    for _, key in ipairs(room.objects) do
        local obj = objects[key]
        if obj and not ctx.excluded[key] and not obj.mentioned
                and obj.location == currentRoomKey then
            if hasActiveSpecialDesc(obj) then
                if obj.specialDescBeforeContents ~= false then
                    firstSpecial[#firstSpecial + 1] = obj
                else
                    secondSpecial[#secondSpecial + 1] = obj
                end
            elseif obj.listed ~= false and not obj.scenery
                   and not obj.remapIn and not obj.remapOn
                   and not (obj.contType == "in" and obj.isOpen) then
                miscItems[#miscItems + 1] = obj
            end
        end
    end

    table.sort(miscItems, function(a, b) return a.name < b.name end)

    local function byOrder(a, b)
        return (a.specialDescOrder or 100) < (b.specialDescOrder or 100)
    end
    table.sort(firstSpecial,  byOrder)
    table.sort(secondSpecial, byOrder)

    local result = {}

    for _, obj in ipairs(firstSpecial) do
        local text = showSpecialDesc(obj)
        if text then result[#result + 1] = text end
    end

    if #miscItems > 0 then
        result[#result + 1] = buildMiscSentence(miscItems)
    end

    for _, obj in ipairs(secondSpecial) do
        local text = showSpecialDesc(obj)
        if text then result[#result + 1] = text end
    end

    -- Stage 4A: Composite objects (remapIn / remapOn) get a dedicated paragraph.
    -- They are excluded from the misc sentence and described here instead.
    for _, key in ipairs(room.objects) do
        local obj = objects[key]
        if obj and not ctx.excluded[key] and not obj.mentioned
                and obj.location == currentRoomKey
                and (obj.remapIn or obj.remapOn) then
            local lines = { "There is a " .. obj.name .. " here." }
            if obj.remapOn then
                local sub = World.getObject(obj.remapOn)
                if sub then
                    local contents = World.contentsOf(sub._key)
                    local names = {}
                    for _, item in ipairs(contents) do
                        if not item.mentioned then
                            names[#names + 1] = item.name
                            item.mentioned = true
                        end
                    end
                    if #names > 0 then
                        lines[#lines + 1] = "On the " .. sub.name .. ": " ..
                                             table.concat(names, ", ") .. "."
                    end
                    sub.mentioned = true
                end
            end
            if obj.remapIn then
                local sub = World.getObject(obj.remapIn)
                if sub then
                    if sub.isOpen then
                        local contents = World.contentsOf(sub._key)
                        local names = {}
                        for _, item in ipairs(contents) do
                            if not item.mentioned then
                                names[#names + 1] = item.name
                                item.mentioned = true
                            end
                        end
                        if #names > 0 then
                            lines[#lines + 1] = "The " .. sub.name .. " is open. It contains: " ..
                                                 table.concat(names, ", ") .. "."
                        else
                            lines[#lines + 1] = "The " .. sub.name .. " is open and empty."
                        end
                    end
                    sub.mentioned = true
                end
            end
            obj.mentioned = true
            result[#result + 1] = table.concat(lines, " ")
        end
    end

    -- Stage 4B: Direct containers in the room (on-surfaces and open in-containers).
    for _, key in ipairs(room.objects) do
        local cont = objects[key]
        if cont and cont.contType and not ctx.excluded[key] and not cont.mentioned
                and cont.location == currentRoomKey then
            if cont.contType == "on" then
                local contents = World.contentsOf(key)
                local names = {}
                for _, item in ipairs(contents) do
                    if not item.mentioned then
                        names[#names + 1] = item.name
                        item.mentioned = true
                    end
                end
                if #names > 0 then
                    result[#result + 1] = "On the " .. cont.name .. ": " ..
                                           table.concat(names, ", ") .. "."
                end
                cont.mentioned = true
            elseif cont.contType == "in" and cont.isOpen then
                local contents = World.contentsOf(key)
                local names = {}
                for _, item in ipairs(contents) do
                    if not item.mentioned then
                        names[#names + 1] = item.name
                        item.mentioned = true
                    end
                end
                if #names > 0 then
                    result[#result + 1] = "The " .. cont.name .. " is open. It contains: " ..
                                           table.concat(names, ", ") .. "."
                else
                    result[#result + 1] = "The " .. cont.name .. " is open and empty."
                end
                cont.mentioned = true
            end
            -- Closed in-containers: skip (in misc list or invisible if listed:false).
        end
    end

    if #result == 0 then return nil end
    return table.concat(result, "\n\n")
end

-- Builds the "Exits: north, south." sentence. All exits are listed,
-- including blocked ones — the player can see a door even if it's locked.
local function listExits(room)
    local dirs = {}
    for dir in pairs(room.exits) do
        dirs[#dirs + 1] = dir
    end
    if #dirs == 0 then return nil end
    table.sort(dirs)
    return "Exits: " .. table.concat(dirs, ", ") .. "."
end

-- The master compositor. Assembles the complete room description:
-- title, body, four-stage object listing, exit list.
function World.describeCurrentRoom()
    local room = rooms[currentRoomKey]

    -- Step 1: Reset mentioned flags on all objects.
    unmentionAll()

    local parts = {}

    -- Step 2: Room title (dark name when unlit).
    if isIlluminated(room) then
        parts[#parts + 1] = room.name
    else
        parts[#parts + 1] = room.darkName or "In the dark"
    end

    -- Step 3: Dark branch — suppress everything except darkDesc.
    if not isIlluminated(room) then
        local darkDesc = room.darkDesc or "It is pitch black; you can't see a thing."
        if type(darkDesc) == "function" then darkDesc = darkDesc(room) end
        parts[#parts + 1] = darkDesc
        room.visited = true
        return table.concat(parts, "\n")
    end

    -- Step 4: Build room context (firstVisit flag + exclude helper).
    local excluded = {}
    local ctx = {
        firstVisit = not room.visited,
        excluded   = excluded,
        exclude    = function(key) excluded[key] = true end,
    }

    -- Step 5: Room description body.
    -- Title and body join with a single newline; subsequent blocks use double.
    local desc = room.description(room, ctx)
    local out = parts[1] .. "\n" .. desc

    -- Steps 6–7: Object listing and exit listing (unless suppressed).
    local suppress
    if type(room.suppressListing) == "function" then
        suppress = room.suppressListing()
    else
        suppress = room.suppressListing
    end
    if not suppress then
        local listing = listContents(room, ctx)
        if listing then out = out .. "\n\n" .. listing end

        local exits = listExits(room)
        if exits then out = out .. "\n\n" .. exits end
    end

    -- Step 8: Mark room as visited.
    room.visited = true

    return out
end

return World
